File: ros_ws/manipulator_disturbance_control/src/test_disturbance_client.cpp
/*
 * test_disturbance_client.cpp
 * 
 * Simple client to test the disturbance service
 * Usage: rosrun manipulator_disturbance_control test_disturbance_client <fx> <fy> <fz>
 * 
 * Author: Manipulator Test Team
 */

#include <ros/ros.h>
#include <manipulator_msgs/ApplyDisturbance.h>
#include <sstream>

int main(int argc, char** argv) {
  ros::init(argc, argv, "test_disturbance_client");
  ros::NodeHandle nh;
  
  // Parse command line arguments
  double fx = 0.0, fy = 0.0, fz = 0.0;
  
  if (argc >= 4) {
    fx = std::atof(argv[1]);
    fy = std::atof(argv[2]);
    fz = std::atof(argv[3]);
  } else {
    ROS_INFO("Usage: rosrun manipulator_disturbance_control test_disturbance_client <fx> <fy> <fz>");
    ROS_INFO("Using default values: fx=50.0 fy=0.0 fz=0.0");
    fx = 50.0;
  }
  
  // Create service client
  ros::ServiceClient client = nh.serviceClient<manipulator_msgs::ApplyDisturbance>(
    "/apply_disturbance"
  );
  
  ROS_INFO("Waiting for /apply_disturbance service...");
  client.waitForExistence();
  
  // Prepare service request
  manipulator_msgs::ApplyDisturbance srv;
  srv.request.force.x = fx;
  srv.request.force.y = fy;
  srv.request.force.z = fz;
  
  ROS_INFO("Calling service with force: [%.2f, %.2f, %.2f] N", fx, fy, fz);
  
  // Call service
  if (client.call(srv)) {
    if (srv.response.success) {
      ROS_INFO("SUCCESS: %s", srv.response.message.c_str());
    } else {
      ROS_ERROR("FAILED: %s", srv.response.message.c_str());
    }
  } else {
    ROS_ERROR("Failed to call service /apply_disturbance");
    return 1;
  }
  
  return 0;
}

File: ros_ws/manipulator_disturbance_control/scripts/test_disturbance_client.py
#!/usr/bin/env python3
"""
test_disturbance_client.py

Python client to test the disturbance service
Usage: rosrun manipulator_disturbance_control test_disturbance_client.py [fx] [fy] [fz]

Author: Manipulator Test Team
"""

import rospy
import sys
from manipulator_msgs.srv import ApplyDisturbance, ApplyDisturbanceRequest

def main():
    rospy.init_node('test_disturbance_client_py')
    
    # Parse command line arguments
    if len(sys.argv) >= 4:
        fx = float(sys.argv[1])
        fy = float(sys.argv[2])
        fz = float(sys.argv[3])
    else:
        rospy.loginfo("Usage: rosrun manipulator_disturbance_control test_disturbance_client.py <fx> <fy> <fz>")
        rospy.loginfo("Using default values: fx=50.0 fy=0.0 fz=0.0")
        fx, fy, fz = 50.0, 0.0, 0.0
    
    # Wait for service
    service_name = '/apply_disturbance'
    rospy.loginfo(f"Waiting for {service_name} service...")
    rospy.wait_for_service(service_name)
    
    try:
        # Create service proxy
        apply_disturbance = rospy.ServiceProxy(service_name, ApplyDisturbance)
        
        # Create request
        req = ApplyDisturbanceRequest()
        req.force.x = fx
        req.force.y = fy
        req.force.z = fz
        
        rospy.loginfo(f"Calling service with force: [{fx:.2f}, {fy:.2f}, {fz:.2f}] N")
        
        # Call service
        response = apply_disturbance(req)
        
        if response.success:
            rospy.loginfo(f"SUCCESS: {response.message}")
        else:
            rospy.logerr(f"FAILED: {response.message}")
            
    except rospy.ServiceException as e:
        rospy.logerr(f"Service call failed: {e}")
        return 1
    
    return 0

if __name__ == '__main__':
    sys.exit(main())
